R/groups.R

```diff
@@ -0,0 +1,205 @@
+#'
+#' @param src The source object
+#' @param groups Either a data frame of groups, or a character vector of group guids
+#'
+#' @return
+#' A tibble with the following columns:
+
+#'   * `group_guid`: The group's GUID
+#'   * `group_name`: The group's name
+#'   * `content_guid`: The content item's GUID
+#'   * `content_name`: The content item's name
+#'   * `content_title`: The content item's title
+#'   * `access_type`: The access type of the content item ("all", "logged_in", or "acl")
+#'   * `role`: The access type that members of the group have to the
+#'     content item, "publisher" or "viewer".
+#'
+#' @examples
+#' \dontrun{
+#' library(connectapi)
+#' client <- connect()
+#'
+#' # Get a data frame of groups
+#' groups <- get_groups(client)
+#'
+#' # Get permissions for a single group by passing in the corresponding row.
+#' get_group_content(client, groups[1, ])
+#' dplyr::filter(groups, name = "research_scientists") %>%
+#'   get_group_content(client, groups = .)
+#'
+#' # Get permissions for all groups by passing in the entire groups data frame.
+#' get_group_content(client, groups)
+#'
+#' # You can also pass in a guid or guids as a character vector.
+#' get_group_content(client, groups$guid[1])
+#' }
+#'
+#' @family groups functions
+#' @export
+#' @importFrom rlang .data
+get_group_content <- function(src, groups) {
+  validate_R6_class(src, "Connect")
+  if (inherits(groups, "data.frame")) {
+    validate_df_ptype(groups, tibble::tibble(
+      guid = NA_character_,
+      name = NA_character_
+    ))
+  } else if (inherits(groups, "character")) {
+    # If a character vector, we assume we are receiving group guids, and call
+    # the endpoint to fetch the group name.
+    groups <- purrr::map_dfr(groups, src$group_details)
+  } else {
+    stop("`groups` must be a data frame or character vector.")
+  }
+
+  purrr::pmap_dfr(
+    dplyr::select(groups, .data$guid, .data$name),
+    get_group_content_impl,
+    src = src
+  )
+}
+
+#' @importFrom rlang .data
+get_group_content_impl <- function(src, guid, name) {
+  validate_R6_class(src, "Connect")
+
+  res <- src$group_content(guid)
+  parsed <- parse_connectapi_typed(res, connectapi_ptypes$group_content)
+
+  dplyr::transmute(parsed,
+    group_guid = guid,
+    group_name = name,
+    .data$content_guid,
+    .data$content_name,
+    .data$content_title,
+    .data$access_type,
+    role = purrr::map_chr(
+      .data$permissions,
+      extract_role,
+      principal_guid = guid
+    )
+  )
+}
+
+# Given the list of permissions for a content item, extract the role for the
+# provided principal_guid
+extract_role <- function(permissions, principal_guid) {
+  matched <- purrr::keep(
+    permissions,
+    ~ .x[["principal_guid"]] == principal_guid
+  )
+  if (length(matched) == 1) {
+    return(matched[[1]][["principal_role"]])
+  } else {
+    stop("Unexpected permissions structure.")
+  }
+  stop(glue::glue("Could not find permissions for \"{principal_guid}\""))
+}
```

Comment:
New functions that comprise the implementations of the `get_group_content()`, factored out slightly for clarity of code and testability.

Notes for reviewers here, which should probably become code comments:

- The exported function, `get_group_content()`, handles input validation and normalization (to allow the use of both GUIDs and data frames), and calls `get_group_content_impl()` across all input groups.
- `get_group_content_impl()` gets the content for a single group from the guid and name. It isn't exported because it is the strictly less flexible core of `get_group_content()`.
- `extract_role()` is factored out for unit tests. Probably also useful elsewhere.